### PR TITLE
[AST] Protocol conformance cleanups

### PR DESCRIPTION
**Pull request by DougGregor**
Various small cleanups for `ProtocolConformance` and its subclasses.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
Whoops, missed a bit. Sorry @swift-ci

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/SILGen/SILGenType.cpp [hunk 1]
```diff
@@ -608,7 +608,9 @@ SILGenModule::emitProtocolWitness(ProtocolConformance *conformance,
     genericEnv = witnessRef.getDecl()->getInnermostDeclContext()
                    ->getGenericEnvironmentOfContext();
 
-    Type concreteTy = conformance->getInterfaceType();
+    auto conformanceDC = conformance->getDeclContext();
+    Type concreteTy =
```
Comment (slavapestov) [inline lib/SILGen/SILGenType.cpp]: Is this ever different from conformanceDC->getSelfInterfaceType()?

**Comment by slavapestov** [comment]
Excellent!

Do you think normal conformances should be written in terms of interface types? In serialization there's a hack to map substitutions out of context and map them back in when serializing conformances, presumably because the normal conformance can form a cycle between a type and its generic environment.

If normal conformances stored interface types the map in/out of context can go away; the only place we should then be serializing archetypes inside substitutions is in the body of a SIL function, but there's no circularity issue there.

File: lib/SILGen/SILGenType.cpp (hunk 1 above)
Comment (DougGregor) [inline lib/SILGen/SILGenType.cpp]: Good point. No, it isn't.

**Comment by DougGregor** [comment]
Yeah, type witnesses, too, could be in terms of interface types rather than contextual types. I do think that would be cleaner.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**